### ISSUE TITLE
With FSA on, analyzer-diagnostics seem super slow to get up to date.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#14993]
I'm trying out the new RC bits.  I've turned FSA on for C# so i can find out about issues and run 'fix all' on them to update code.  For example, I've gone and run the "Use throw expression" fix on the EditorFeatures project.

I have roslyn.sln opened from open, and my master pointer is at af7d1c7ea631e5543167146e172268f269d1f0da

The diagnostics we've reported for these never seem to be cleared from the error list window until i go navigate to the file that was changed.  I find this surprising given that edits occurred in these files *and* these files are part of the project i currently have open.  I would have expected the diagnostic engine to go and update these files quickly to help clear out stale messages.

--

Note: this makes the fix-all-in-XXX experience quite poor.  You want to run fix-all, and see htose errors go away in a reasonable amount of time.  I've had my IDE running for several minutes after this, and they *never* seem to clear.  I don't know if they'll eventually go away.  But right now it either isn't happening, or is happening so slowly that the experience is really bad.

**Comment by CyrusNajmabadi** [comment]
Tagging @srivatsn @heejaechang @mavasani   For their thoughts on what's going on here.

**Comment by CyrusNajmabadi** [comment]
For example, these messages:

![image](https://cloud.githubusercontent.com/assets/4564579/20020965/1e240880-a291-11e6-88aa-86e2f1fbf609.png)

These issues were fixed by running fix-all at least 10+ minutes ago.  They go away if i visit the files they're referring to.  But nothing else seems to clear them out automatically at any sort of reasonable rate.  We should probably front-load analysis of edited documents to make sure our errors are fresh.

**Comment by CyrusNajmabadi** [comment]
Interestingly: the OOP process is not using CPU in the background.  Leading me to think it's done analyzing.  So, AFAICT, these errors simply will not go away unless i open the file.  That experience is clearly broken.

**Comment by heejaechang** [comment]
@CyrusNajmabadi do you see ... in error list? is devenv still processing? all diagnostics that doesnt belong to open file has very low priority.

**Comment by CyrusNajmabadi** [comment]
> all diagnostics that doesnt belong to open file has very low priority.

This is not a good user experience.  Files the user has directly edited, whether they are open or not, should be updated quickly so that stale diagnostics don't last.

**Comment by heejaechang** [comment]
if you see ... in error list and devenv using CPU then it is processing something else. when file changed, diagnostic is not the only one that need to reprocess those files. if many files are changed, devenv probably reprocessing whole solution, once that is done it will move to low priority ones.

**Comment by CyrusNajmabadi** [comment]
You're explaining the implementation.  I'm simply saying that the experience is bad :)   The explanation doesn't make the experience less bad :)

**Comment by heejaechang** [comment]
well, I am setting your expectation. renaming overload in syntaxnode to something else takes rename more than 1 hour. experience is bad. but not way out of expectation.

**Comment by heejaechang** [comment]
but regardless of that, whether something is broken or slow is different thing. if error list doesn't show ... and errors doesnt go away then, it is broken, if it still shows ... then it is slow.

**Comment by CyrusNajmabadi** [comment]
No "... "

![image](https://cloud.githubusercontent.com/assets/4564579/20021547/7d167c40-a293-11e6-8b28-01b0243511a5.png)

**Comment by heejaechang** [comment]
then, something is broken.

**Comment by CyrusNajmabadi** [comment]
When i visit that 3rd item, it then disappears.

**Comment by heejaechang** [comment]
how can I make use "throw" expression info to show up?

**Comment by CyrusNajmabadi** [comment]
Enable this feature:

![image](https://cloud.githubusercontent.com/assets/4564579/20021765/9565d696-a294-11e6-9706-71a5d26e62fc.png)

**Comment by heejaechang** [comment]
@CyrusNajmabadi so when you apply fix, fix part worked? I am trying to find out which part is broken. if fix worked then solution crawler seems somehow not picking up ones that changed.

**Comment by CyrusNajmabadi** [comment]
Yes.  All the files are fixed up.  But the error list doesn't clear properly.  When you finally go *visit* the file, you'll see that the fix happened.  And at *that* point, the suggestion gets removed from the error list.

**Comment by heejaechang** [comment]
@CyrusNajmabadi is it RC bug because it happens only with OOP or is it happening regardless? if it is OOP issue, fix more likely won't worked.

@srivatsn is this something we want to bring in RC?

**Comment by CyrusNajmabadi** [comment]
it's an RC bug because it's a bad experience with the RC build :)

**Comment by mavasani** [comment]
I think the right fix lies in the code fix engine - solution crawler will be slow in refreshing these diagnostics unless it is told to prioritize re-analysis for the changed scope. We faced similar issue when adding support for bulk suppress diagnostics from error list and [this](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/TableDataSource/Suppression/VisualStudioSuppressionFixService.cs,361) code was added after the suppression fix all is executed to refresh the diagnostics quicker. I think same approach should work in code fix engine.

**Comment by heejaechang** [comment]
that will work too. but what Cyrus experiencing is solution crawler is done running but errors are still in error list.

**Comment by CyrusNajmabadi** [comment]
That's certainly a possiblity.  However, IMO, it makes sense to me that edited files *always* get bumped up in priority.  They're being manipulated by teh user in some way (if not directly in an open file).  So they should get preference over files that are not being touched at all.

**Comment by heejaechang** [comment]
well... if error happen to be semantic ones like missing or error type, just making changed files higher priority won't make things clear faster. sure it will make 1 case better but not all cases..

**Comment by CyrusNajmabadi** [comment]
1 case better is better than no cases better :)

**Comment by srivatsn** [comment]
Let's figure out the right fix and fix it for RC2.

**Comment by heejaechang** [comment]
what I am saying is if you happen to use some other code fix that caused semantic propagation, error would have still removed slower than you expected with even more random behavior such as that some errors go away quickly and others left there long time than you expected. still bad experience.

**Comment by CyrusNajmabadi** [comment]
I do understand that :)

**Comment by heejaechang** [comment]
hmm I have been trying to repro. not reproing. could be some race in solution crawler...
